man/dmbeat.1
.TH "DMBEAT" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmbeat \(en sends status messages to an RPC server
.SH SYNOPSIS
\f[B]dmbeat\f[R] \-\-help
.PP
\f[B]dmbeat\f[R] \-\-version
.PP
\f[B]dmbeat\f[R] \-\-\f[B]node\f[R] \f[I]id\f[R] \-\-\f[B]host\f[R]
\f[I]host\f[R] [\-\-\f[B]port\f[R] \f[I]port\f[R]] [\-\-\f[B]tls\f[R]]
[\-\-\f[B]username\f[R] \f[I]user\f[R]] [\-\-\f[B]password\f[R]
\f[I]password\f[R]] [\-\-\f[B]count\f[R] \f[I]n\f[R]]
[\-\-\f[B]interval\f[R] \f[I]seconds\f[R]] [\-\-\f[B]compression\f[R]
\f[I]name\f[R]] [\-\-\f[B]logger\f[R] \f[I]name\f[R]]
[\-\-\f[B]debug\f[R]] [\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmbeat\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
This program is a heartbeat emitter that sends status messages to a
remote DMPACK HTTP\-RPC API.
The status messages include timestamp, system uptime, and last
connection error.
The server may inspect this data to check if a client is still running
and has network access.
The RPC endpoint is expected at
\f[CR][http|https]://<host>:<port>/api/v1/beat\f[R].
The heartbeat is transmitted in Zstandard\-compressed Fortran 95
Namelist format.
.PP
Passing the server credentials via the command\-line arguments
\f[CR]\-\-username\f[R] and \f[CR]\-\-password\f[R] is insecure on
multi\-user operating systems and only recommended for testing.
.SH OPTIONS
.TP
\f[B]\-\-compression\f[R], \f[B]\-x\f[R] \f[I]name\f[R]
Compression library to use for payload data (\f[CR]none\f[R],
\f[CR]zlib\f[R], \f[CR]zstd\f[R]).
The default is \f[CR]zstd\f[R] (Z standard).
Select \f[CR]zlib\f[R] for deflate compression.
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
Path to Lua\-based configuration file.
.TP
\f[B]\-\-count\f[R], \f[B]\-C\f[R] \f[I]n\f[R]
Maximum number of heartbeats to send.
The default is 0 (unlimited).
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-host\f[R], \f[B]\-H\f[R] \f[I]host\f[R]
IP or FQDN of RPC host (for example, \f[CR]127.0.0.1\f[R] or
\f[CR]example.com\f[R]).
.TP
\f[B]\-\-interval\f[R], \f[B]\-I\f[R] \f[I]seconds\f[R]
Emit interval in seconds (default is 60 seconds).
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of instance and table in given configuration file (default is
\f[CR]dmbeat\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Sensor node id.
.TP
\f[B]\-\-password\f[R], \f[B]\-P\f[R] \f[I]password\f[R]
RPC API password.
Be aware that passing the credentials via command\-line arguments is
insecure and only recommended for testing.
.TP
\f[B]\-\-port\f[R], \f[B]\-q\f[R] \f[I]port\f[R]
Port of RPC API server.
The default port is 0 (automatic selection depending on protocol).
.TP
\f[B]\-\-tls\f[R], \f[B]\-E\f[R]
Use TLS encryption.
.TP
\f[B]\-\-username\f[R], \f[B]\-U\f[R] \f[I]user\f[R]
RPC API user name.
If set, implies HTTP Basic Auth.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Beat has been transmitted.
.TP
\f[B]1\f[R]
Failure.
Beat transmission failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Send a single heartbeat to a \f[I]dmapi(1)\f[R] process on localhost:
.IP
.EX
$ dmbeat \-N dummy\-node \-H 127.0.0.1 \-C 1 \-V
.EE
.SH SEE ALSO
\f[I]dmapi(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmbeat.f90
! dmbeat.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmbeat
    !! Heartbeat emitter, sends status messages to the RPC API.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmbeat'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    integer, parameter :: APP_HOST_LEN     = 256 !! Max. length of host name.
    integer, parameter :: APP_USERNAME_LEN = 256 !! Max. length of user name.
    integer, parameter :: APP_PASSWORD_LEN = 256 !! Max. length of password.

    type :: app_type
        !! Application settings.
        character(ID_LEN)           :: name             = APP_NAME    !! Name of instance/configuration.
        character(FILE_PATH_LEN)    :: config           = ' '         !! Path to configuration file.
        character(LOGGER_NAME_LEN)  :: logger           = ' '         !! Name of logger (name implies IPC).
        character(NODE_ID_LEN)      :: node_id          = ' '         !! Sensor node id (required).
        character(APP_HOST_LEN)     :: host             = ' '         !! IP or FQDN of API (`127.0.0.1`, `example.com`).
        integer                     :: port             = 0           !! API port (set to 0 for protocol default).
        logical                     :: tls              = .false.     !! TLS encryption.
        character(APP_USERNAME_LEN) :: username         = ' '         !! HTTP Basic Auth user name.
        character(APP_PASSWORD_LEN) :: password         = ' '         !! HTTP Basic Auth password.
        character(Z_TYPE_NAME_LEN)  :: compression_name = 'zstd'      !! Compression library (`none`, `zlib`, `zstd`).
        integer                     :: compression      = Z_TYPE_NONE !! Compression type (`Z_TYPE_*`).
        integer                     :: count            = 0           !! Maximum number of heartbeats to send (0 means unlimited).
        integer                     :: interval         = 60          !! Emit interval in seconds (>= 0).
        logical                     :: debug            = .false.     !! Forward debug messages via IPC.
        logical                     :: verbose          = .false.     !! Print debug messages to stderr.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    call dm_init()

    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward DEBUG messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call dm_posix_signal_register(signal_callback)

    rc = run(app)
    call shutdown(rc)
contains
    integer function run(app) result(rc)
        !! Runs main loop to emit heartbeats.
        type(app_type), intent(inout) :: app !! App type.

        character(BEAT_CLIENT_LEN) :: client
        character(LOG_MESSAGE_LEN) :: message
        character(:), allocatable  :: url

        integer     :: iter, rc_last, stat
        integer     :: msec, sec
        integer(i8) :: uptime
        logical     :: has_api_status

        type(api_status_type)   :: api_status
        type(beat_type)         :: beat
        type(rpc_request_type)  :: request
        type(rpc_response_type) :: response
        type(timer_type)        :: timer

        rc = dm_rpc_init()

        if (dm_is_error(rc)) then
            call logger%error('failed to initialize libcurl', error=rc)
            return
        end if

        call logger%info('started ' // APP_NAME)
        call logger%debug('beat transmission interval: ' // dm_itoa(app%interval))

        if (app%compression == Z_TYPE_NONE) then
            call logger%debug('compression is disabled')
        else
            call logger%debug(dm_z_type_name(app%compression) // ' compression is enabled')
        end if

        client  = dm_version_to_string(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH, library=.true.)
        rc_last = E_NONE

        ! Initialise heartbeat.
        call dm_beat_set(beat, node_id=app%node_id, client=client, interval=app%interval)

        ! Create URL of RPC service.
        url = dm_rpc_url(host=app%host, port=app%port, endpoint=RPC_ROUTE_BEAT, tls=app%tls)

        rc = E_INVALID
        if (.not. dm_string_has(url)) return

        call logger%debug('sending beats to API endpoint ' // url)
        iter = 1

        emit_loop: do
            call dm_timer_start(timer)
            if (app%count > 0) call logger%debug('starting transmission ' // dm_itoa(iter) // '/' // dm_itoa(app%count))

            ! Update heartbeat attributes.
            call dm_posix_uptime(uptime, stat)
            call dm_beat_set(beat, time_sent=dm_time_now(), error=rc_last, uptime=int(uptime))

            ! Send RPC request to API, use compression if available.
            rc = dm_rpc_post(request, response, beat, url=url, username=app%username, password=app%password, user_agent=client, compression=app%compression)
            if (dm_is_error(rc)) call logger%debug('failed to send beat to host ' // app%host, error=rc)

            ! Read API status response from payload.
            has_api_status = .false.

            if (response%content_type == MIME_TEXT) then
                stat = dm_api_status_from_string(response%payload, api_status)
                has_api_status = dm_is_ok(stat)
            end if

            if (response%code > 0) then
                call logger%debug('server answered with status HTTP ' // dm_itoa(response%code) // ' ' // dm_http_status_string(response%code))
            end if

            ! Log the HTTP response code.
            select case (response%code)
                case (HTTP_NONE)
                    rc = E_RPC_CONNECT
                    call logger%warning('connection to host ' // trim(app%host) // ' failed: ' // response%error_message, error=rc)

                case (HTTP_CREATED)
                    rc = E_NONE
                    call logger%debug('beat accepted by host ' // app%host, error=rc)

                case (HTTP_UNAUTHORIZED)
                    rc = E_RPC_AUTH
                    message = 'unauthorized access on host ' // app%host
                    if (has_api_status) message = trim(message) //  ': ' // api_status%message
                    call logger%error(message, error=rc)

                case (HTTP_INTERNAL_SERVER_ERROR)
                    rc = E_RPC_SERVER
                    call logger%error('internal server error on host ' // app%host, error=rc)

                case (HTTP_BAD_GATEWAY)
                    rc = E_RPC_CONNECT
                    call logger%error('bad gateway on host ' // app%host, error=rc)

                case default
                    rc = E_RPC_API
                    write (message, '("API call to host ", a, " failed (HTTP ", i0, ")")') trim(app%host), response%code

                    if (has_api_status) then
                        rc = api_status%error
                        message = trim(message) // ': ' // api_status%message
                    end if

                    call logger%warning(message, error=rc)
            end select

            rc_last = rc
            call dm_rpc_reset(response)

            if (app%count > 0) then
                call logger%debug('finished transmission ' // dm_itoa(iter) // '/' // dm_itoa(app%count))
                iter = iter + 1
                if (iter > app%count) exit emit_loop
            end if

            call dm_timer_stop(timer)
            msec = max(0, int(1000 * (app%interval - dm_timer_result(timer))))
            sec  = dm_msec_to_sec(msec)
            call logger%debug('next beat in ' // dm_itoa(sec) // ' sec')
            call dm_msleep(msec)
        end do emit_loop

        call dm_rpc_destroy(request)
        call dm_rpc_destroy(response)

        rc = E_NONE
        call logger%debug('finished transmission')
    end function run

    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))
        call dm_rpc_shutdown()

        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app !! App type.

        type(arg_parser_class) :: parser

        call parser%add('name',        short='n', type=ARG_TYPE_ID)      ! -n, --name <id>
        call parser%add('config',      short='c', type=ARG_TYPE_FILE)    ! -c, --config <path>
        call parser%add('logger',      short='l', type=ARG_TYPE_ID)      ! -l, --logger <id>
        call parser%add('node',        short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('host',        short='H', type=ARG_TYPE_STRING)  ! -H, --host <string>
        call parser%add('port',        short='q', type=ARG_TYPE_INTEGER) ! -q, --port <n>
        call parser%add('tls',         short='E', type=ARG_TYPE_LOGICAL) ! -E, --tls
        call parser%add('username',    short='U', type=ARG_TYPE_STRING)  ! -U, --username <string>
        call parser%add('password',    short='P', type=ARG_TYPE_STRING)  ! -P, --password <string>
        call parser%add('compression', short='x', type=ARG_TYPE_STRING)  ! -x, --compression <name>
        call parser%add('count',       short='C', type=ARG_TYPE_INTEGER) ! -C, --count <n>
        call parser%add('interval',    short='I', type=ARG_TYPE_INTEGER) ! -I, --interval <n>
        call parser%add('debug',       short='D', type=ARG_TYPE_LOGICAL) ! -D, --debug
        call parser%add('verbose',     short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Get all other arguments.
        call parser%get('logger',      app%logger)
        call parser%get('node',        app%node_id)
        call parser%get('host',        app%host)
        call parser%get('port',        app%port)
        call parser%get('tls',         app%tls)
        call parser%get('username',    app%username)
        call parser%get('password',    app%password)
        call parser%get('compression', app%compression_name)
        call parser%get('count',       app%count)
        call parser%get('interval',    app%interval)
        call parser%get('debug',       app%debug)
        call parser%get('verbose',     app%verbose)

        if (dm_string_has(app%compression_name)) app%compression = dm_z_type_from_name(app%compression_name)

        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from file.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('logger',      app%logger)
            call config%get('node',        app%node_id)
            call config%get('host',        app%host)
            call config%get('port',        app%port)
            call config%get('tls',         app%tls)
            call config%get('username',    app%username)
            call config%get('password',    app%password)
            call config%get('compression', app%compression_name)
            call config%get('count',       app%count)
            call config%get('interval',    app%interval)
            call config%get('debug',       app%debug)
            call config%get('verbose',     app%verbose)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger name')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid or missing node id')
            return
        end if

        if (.not. dm_string_has(app%host)) then
            call dm_error_out(rc, 'invalid or missing host')
            return
        end if

        if (app%count < 0) then
            call dm_error_out(rc, 'invalid count')
            return
        end if

        if (app%interval < 0) then
            call dm_error_out(rc, 'invalid interval')
            return
        end if

        if (.not. dm_z_is_valid(app%compression)) then
            call dm_error_out(rc, 'invalid compression')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(signum) bind(c)
        !! C-interoperable signal handler that stops the program.
        integer(c_int), intent(in), value :: signum !! Signal number.

        call logger%debug('exit on signal ' // dm_posix_signal_name(signum))
        call shutdown(E_NONE)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a, 2(1x, a))', dm_rpc_version(), dm_lua_version(.true.), dm_zstd_version(.true.)
    end subroutine version_callback
end program dmbeat
